### PR TITLE
Improve query performance for ClickHouse on S3

### PR DESCRIPTION
**Pull request by fastio**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):

- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

Reduce the amount of data for s3 engine when executing query.

Detailed description / Documentation draft:

First, When using ClickHouse to query the data on the object store (S3, COS), in some cases a large amount of unnecessary data will be read, resulting in poor query performance. For example, when the required data is close to the end of a large file when SQL is executed, ClickHouse will start to read the data from the head of the large file until it obtains the required data.

Simply put, SeekAvoidingReadBuffer is currently used to implement the seek operation through ignore. ReadIndirectBufferFromS3::ignore reads data from the head until the desired position.

```
Call path:
SeekAvoidingReadBuffer::seek -> ReadBuffer::ignore -> ReadBufferFromS3::nextImpl -> ReadBufferFromIStream::nextImpl
```

Second, ClickHouse can accurately know the offset and size of the data in the S3 file for the query. We tell the object storage system the accurate offset and size by SetRange function, which will save the preparation time of the object storage backend and obtain better time.

In Tencent COS, if the range for reading data is not clearly provided, the return time of GetObject will become longer.

### CODE REVIEW

**Comment by MaxWk** [comment]
it's very useful

**Comment by excitoon** [comment]
I've measures read performance for different ranges (for Moscow<-London and same region). For small pieces of data it is more efficient to read from the start. Specifying end range effects in minor slowdown.

Figure_1
